storage_controller/src/service.rs

```diff
@@ -4312,6 +4312,13 @@ impl Service {
                         continue;
                     }
 
+                    if !new_nodes.values().any(Node::is_available) {
+                        // Special case for when all nodes are unavailable: there is no point
+                        // trying to reschedule since there's nowhere else to go. Without this
+                        // branch we incorrectly detach tenants in response to node unavailability.
+                        continue;
```

Comment:
If we leave tenant shards with their IntentState::attached set to an offline node, and then all the _other_ nodes come online, there isn't currently anything in the background loop that will switch their attachment to an available node.

However, this isn't a regression in this PR, because it was already the case if we set all nodes to just have secondary locations.

Reply:
Right. I'd say the new behaviour is preferable. We'll have to fix the issue you are alluding to in any case.